_posts/2021-12-08-kafkasql-storage-and-security.md

```diff
@@ -0,0 +1,471 @@
+
+```json
+{
+  "topic": "kafkasql-journal",
+  "partition": 0,
+  "offset": 1,
+  "tstype": "create",
+  "ts": 1638894723320,
+  "headers": [
+    "req",
+    "9dff3fb8-7a1d-41b0-9083-684d9fbfa619"
+  ],
+  "key": "\t{\"tenantId\":\"_\"}",
+  "payload": "\t{\"action\":\"CREATE\"}"
+}
+```
+
+And the same message after the bugfix:
+
+```json
+{
+  "topic": "kafkasql-journal",
+  "partition": 0,
+  "offset": 1,
+  "tstype": "create",
+  "ts": 1638894723320,
+  "headers": [
+    "req",
+    "9dff3fb8-7a1d-41b0-9083-684d9fbfa619"
+  ],
+  "key": "\t{\"tenantId\":\"_\",\"uuid\":\"6875bbc9-92fe-4572-accc-7dab2bdcd84e\"}",
+  "payload": "\t{\"action\":\"CREATE\"}"
+}
+```
+
+# Temporary Resolution for Older Apicurio Registry Versions
+
+This issue has already been fixed in Apicurio Registry `2.1.4.Final`. If you are not able to upgrade, follow this procedure:
+
+In order to make sure you are not affected by this bug, you need to disable log compaction for the `kafkasql-journal` topic
+in your cluster. First, check if it is enabled:
+
+```shell
+bin/kafka-configs.sh --describe --bootstrap-server localhost:9092 --topic kafkasql-journal 
+```
+
+If you see `cleanup.policy=compact`, you are affected. To disable log compaction, you need to switch to the `delete`
+cleanup policy. However, in case of Apicurio Registry, deleting old messages may result in data corruption. 
+Therefore, you also have to set `retention.bytes=-1` and `retention.ms=-1` (or their server-wide equivalents 
+`log.retention.bytes` and `log.retention.ms`). For example:
+
+```shell
+bin/kafka-configs.sh --alter --bootstrap-server localhost:9092 --topic kafkasql-journal \
+  --add-config cleanup.policy=delete \
+  --add-config retention.bytes=-1 \
+  --add-config retention.ms=-1
+```
+
+If you are using Strimzi, you can apply these settings in the `config` section of the `KafkaTopic` or `Kafka` CRD:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaTopic
+metadata:
+  name: kafkasql-journal
+  labels:
+    strimzi.io/cluster: my-cluster
+  namespace: registry-example-kafkasql
+spec:
+  partitions: 2
+  replicas: 1
+  config:
+    cleanup.policy: delete
+    retention.bytes: -1
+    retention.ms: -1
+```
+*Strimzi KafkaTopic CRD example*
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+  namespace: registry-example-kafkasql
+spec:
+  kafka:
+    config:
+      # ...
+      log.cleanup.policy: delete
+      log.retention.bytes: -1
+      log.retention.ms: -1
+    version: 2.7.0
+    # ...
+```
+*Strimzi Kafka CRD example*
+
+However, this configuration change does not fix the existing corrupted data in the database. 
+Luckily, we can use the export-import feature of the Apicurio Registry to force a reset of the database sequences,
+so the error no longer occurs. This will keep your existing data intact.
```

Comment:
Have we confirmed if this works?  I think this might have been my responsibility to test and if so I have not done it.

Reply:
I have confirmed it.